driver-core/src/main/com/mongodb/internal/connection/CompositeByteBuf.java

```diff
@@ -315,6 +321,11 @@ public void release() {
             referenceCount.incrementAndGet();
             throw new IllegalStateException("Attempted to decrement the reference count below 0");
         }
+        components.forEach(c -> c.buffer.release());
+        if (referenceCount.get() == 0) {
+            Assertions.assertTrue(components.stream().noneMatch(c -> c.buffer.getReferenceCount() > 0),
+                    "All component buffers should have reference count 0 when CompositeByteBuf is fully released, but some still have references.");
```

Comment:
This assertion combined with the fact that `release`/`retain` update `referenceCount` together with the reference count of each `component.buffer`, suggests to me that immediately after (in program order) constructing a `CompositeByteBuf` instance, the reference count of each `component.buffer` must be equal to `referenceCount`, which is 1. If this reasoning is correct, we should introduce such an assertion to the constructors of `CompositeByteBuf`.

Reply:
Done